src/backend/opencl/kernel/moments.hpp

```diff
@@ -0,0 +1,103 @@
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+#include <kernel_headers/moments.hpp>
+#include <program.hpp>
+#include <traits.hpp>
+#include <string>
+#include <mutex>
+#include <map>
+#include <dispatch.hpp>
+#include <Param.hpp>
+#include <debug_opencl.hpp>
+#include <type_util.hpp>
+#include <math.hpp>
+#include "config.hpp"
+
+using cl::Buffer;
+using cl::Program;
+using cl::Kernel;
+using cl::make_kernel;
+using cl::EnqueueArgs;
+using cl::NDRange;
+using std::string;
+
+namespace opencl
+{
+    namespace kernel
+    {
+        static const int THREADS = 128;
+
+        ///////////////////////////////////////////////////////////////////////////
+        // Wrapper functions
+        ///////////////////////////////////////////////////////////////////////////
+        template <typename T, af_moment moment>
+        void moments(Param out, const Param in)
+        {
+            try {
+                static std::once_flag compileFlags[DeviceManager::MAX_DEVICES];
+                static std::map<int, Program*>  momentsProgs;
+                static std::map<int, Kernel*> momentsKernels;
+
+                int device = getActiveDeviceId();
+
+                std::call_once( compileFlags[device], [device] () {
+                    std::ostringstream options;
+                    options << " -D T="        << dtype_traits<T>::getName();
+
+                    if (std::is_same<T, double>::value ||
+                        std::is_same<T, cdouble>::value) {
+                        options << " -D USE_DOUBLE";
+                    }
+
+                    switch(moment) {
+                        case M00:   options << " -D MOMENT=M00";
+                            break;
+                        case M01:   options << " -D MOMENT=M01";
+                            break;
+                        case M10:   options << " -D MOMENT=M10";
+                            break;
+                        case M11:   options << " -D MOMENT=M11";
+                            break;
+                        default:
+                            break;
+                    }
+
+                    Program prog;
+                    buildProgram(prog, moments_cl, moments_cl_len, options.str());
+                    momentsProgs[device] = new Program(prog);
+
+                    momentsKernels[device] = new Kernel(*momentsProgs[device], "moments_kernel");
+                });
+
+
+                auto momentsp = make_kernel<Buffer, const KParam, const Buffer, const KParam,
+                                       const int, const int>
+                                      (*momentsKernels[device]);
+
+                NDRange local(THREADS, 1, 1);
+                dim_t blocksPerMat = divup(in.info.dims[0], local[0]);
```

Comment:
The variable name is a misnomer. it should technically be blocksPerRow / blocksPerX